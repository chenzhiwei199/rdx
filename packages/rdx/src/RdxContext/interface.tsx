import {
  Point,
  ISnapShotTrigger,
  IStatusInfo,
  TASK_PROCESS_TYPE,
  BasePoint,
} from '@czwcode/task-queue';
import { TaskStatus, ShareContextClass } from './shareContext';
import { Base } from './core';
import { IRdxTask } from '../global';

export type MapObject<T> = { [key: string]: T | null };

export interface RdxContextProps<GModel> {
  name?: string;
  children: React.ReactNode;
  context?: React.Context<ShareContextClass>;
  withRef?: React.MutableRefObject<ShareContextClass>;
  initializeState?: MapObject<GModel>;
  onChange?: (state: MapObject<GModel>) => void;
  createStore?: (data: any) => Base<GModel>;
  visualStatePlugins?: React.ReactNode | React.ReactNode[];
  // 依赖数据池
}

export interface INIT_INFO {
  task: Point[];
}
export type ProcessGraphContent = ISnapShotTrigger | IStatusInfo;
export type ProcessType = PROCESS_GRAPH_TYPE | TASK_PROCESS_TYPE;
export interface Process {
  type: ProcessType;
  content: ProcessGraphContent;
}
export enum PROCESS_GRAPH_TYPE {
  INIT = 'INIT',
  TASK_CHANGE = 'TASK_CHANGE',
}

export enum ActionType {
  Update = 'update',
  Remove = 'remove',
  Merge = 'merge',
}
export enum TargetType {
  TasksMap = 'tasks',
  TaskState = 'taskState',
  Trigger = 'trigger',
  CustomAction = 'customAction',
  TaskStatus = 'taskStatus',
  CancelMap = 'cancelMap',
}
export interface Action<GModel> {
  type?: ActionType;
  targetType: TargetType;
  payload?:
    | {
        key: string;
        value: IRdxTask<GModel> | TaskStatus | GModel | (() => void) | null;
      }
    | { points: BasePoint[]; refresh: boolean; executeTask: boolean }
    | { id: string; customAction: any };
}
