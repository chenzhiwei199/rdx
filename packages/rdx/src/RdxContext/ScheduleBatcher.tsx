import React, { useEffect, useState, useContext } from 'react';
import { ShareContextInstance, ShareContextClass } from './shareContext';
import { RdxNode } from '../RdxValues';
import logger from '../utils/log';

const ScheduleBatcher = (props: { setNotifyBatcherOfChange: any }) => {
  const storeRef = useContext<ShareContextClass<any, any>>(
    ShareContextInstance
  );
  const [_, setState] = useState([]);
  props.setNotifyBatcherOfChange(() => setState({} as any));
  useEffect(() => {
    if (storeRef.willNotifyQueue.size > 0) {
      // const depsRdxNodes = new Set<RdxNode<any>>();
      // // 收集相关节点依赖
      // for (let point of storeRef.willNotifyQueue.values()) {
      //   const deps = storeRef.getDeps(point.key);
      //   for (let dep of deps) {
      //     if (dep instanceof RdxNode) {
      //       depsRdxNodes.add(dep);
      //     }
      //   }
      // }
      // // 添加相关节点依赖
      // for (let atom of depsRdxNodes) {
      //   // 不存在才添加
      //   if (!storeRef.getTaskMap(atom.getId())) {
      //     storeRef.addOrUpdateTask(atom.getId(), atom.load(storeRef), false);
      //   }
      // }
      logger.info('Schedule Batcher', Array.from(storeRef.willNotifyQueue));
      storeRef.batchTriggerSchedule(Array.from(storeRef.willNotifyQueue));
    }
    storeRef.willNotifyQueue.clear();
  });

  return null;
};

export default ScheduleBatcher;
