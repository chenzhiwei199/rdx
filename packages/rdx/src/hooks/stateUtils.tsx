export function waitForNone() {}

export function waitForAll<T>(info ) {

}
export function waitForAny() {}

export function noWait() {}
