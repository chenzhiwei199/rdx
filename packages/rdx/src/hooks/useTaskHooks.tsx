import { IRdxViewBase } from '../global';

export type IRdxStateHook<GModel> = IRdxViewBase<GModel> & { id: string };

export function useDefaultValue<IModel, IRelyModel>() {}
