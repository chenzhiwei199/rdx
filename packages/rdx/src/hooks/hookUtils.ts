import { useRef, useEffect, useState } from "react";
import { ShareContextClass } from "../RdxContext/shareContext";
import { StateUpdateType } from "../global";

export function useMount() {
  const mount = useRef(false);
  useEffect(() => {
    mount.current = true;
  }, []);
  return mount;
}

export const useForceUpdate = () => {
  const [state, setState] = useState(1);
  return () => {
    setState((state) => state + 1);
  };
};


export function createStore(): [() => any, (v: any) => void] {
  const storeRef = useRef(null);
  return [
    () => storeRef.current,
    (v) => {
      storeRef.current = v;
    },
  ];
}

export function useGlobalStateUpdate(context: ShareContextClass,) {
  const forceUpdate = useForceUpdate();
  useEffect(() => {
    context.getEventEmitter().on(StateUpdateType.GlobalState, () => {
      forceUpdate();
    });
    return () => {
      context.getEventEmitter().off(StateUpdateType.GlobalState);
    };
  }, []);
}
