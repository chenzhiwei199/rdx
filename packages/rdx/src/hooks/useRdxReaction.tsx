import { useContext, useRef } from 'react';
import { ShareContextInstance } from '../RdxContext/shareContext';
import { DataContext, IRdxViewBase, IRdxView } from '../global';
import { useTaskBinding } from './useTaskHooks';
import { createBaseContext, createMutators } from '../utils';

export type IRdxStateHook<IModel, IRelyModel, IAction> = IRdxViewBase<
  IModel,
  IRelyModel,
  IAction
> & { id: string };
export type IRdxPreviewHook<IModel, IRelyModel, IAction> = IRdxViewBase<
  IModel,
  IRelyModel,
  IAction
>;

//
export function useRdxState<IModel, IRelyModel, IAction>(
  props: IRdxStateHook<IModel, IRelyModel, IAction>
): [IModel, (v: IModel) => void] {
  const context = useRdxContext(props);
  return [context.value, context.next];
}

export function useRdxContext<IModel, IRelyModel, IAction>(
  props: IRdxStateHook<IModel, IRelyModel, IAction>
) {
  const context = useContext(ShareContextInstance);
  useTaskBinding<IModel, IRelyModel, IAction>({ ...props, context });
  const data: DataContext<IModel, IRelyModel> = {
    ...createBaseContext(props.id, context, props),
    ...createMutators(props.id, context),
  };
  return data;
}

let reactionPreviewId = 0;
export function useRdxContextAutoId<IModel, IRelyModel, IAction>(
  props: IRdxPreviewHook<IModel, IRelyModel, IAction>
) {
  const uniqueId = useRef('reaction-' + reactionPreviewId++);
  return useRdxContext({ ...props, id: uniqueId.current });
}
