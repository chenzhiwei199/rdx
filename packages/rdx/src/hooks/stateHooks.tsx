import {
  Status,
  IRdxAnyDeps,
  DataContext,
  IRdxTask,
  StateUpdateType,
} from '../global';
import { RdxNode } from '../RdxValues/base';
import {
  IRdxAtomNode,
  RdxAtomNode,
  IRdxWatcherNode,
  RdxWatcherNode,
} from '../RdxValues';
import { useRef, useMemo, useEffect } from 'react';
import { useRdxStateContext, DefaultContext } from '../RdxContext/shareContext';
import { useForceUpdate, useGlobalStateUpdate } from './hookUtils';
import {
  getDepId,
  createBaseContext,
  createMutators,
} from '../utils/taskUtils';
import { ShareContextClass, TNext } from '..';

export function createRdxHooks(provider = DefaultContext) {
  function getId<GModel>(props: IRdxTask<GModel> | RdxNode<GModel>) {
    if (props instanceof RdxNode) {
      return props.getId();
    } else {
      return props.id;
    }
  }

  function useStateUpdate(id: string, type: StateUpdateType) {
    const context = useRdxStateContext(provider);
    const forceUpdate = useForceUpdate();
    useEffect(() => {
      const eventKey = id + '----' + type;
      context.getEventEmitter().on(eventKey, () => {
        forceUpdate();
      });
      return () => {
        context.getEventEmitter().off(eventKey);
      };
    }, []);
  }

  function useTaskBinding<GModel>(props: RdxNode<GModel>) {
    // 1. 初始化的时候根据atom 默认值类型判断是否是异步任务，判断是否要重跑任务
    // 2. 根据get是否是异步任务，判断是否需要重跑任务
    const context = useRdxStateContext(provider);
    if (!context.getTaskStateById) {
      throw new Error('使用rdx的组件必须为rdxContext子孙组件');
    }
    // 注册节点
    useMemo(() => {
      if (!context.hasTask(props.getId())) {
        props.load(context);
      }
    }, [props.getId()]);

    useEffect(() => {
      return () => {
        //  卸载节点，检查依赖&warning
        // !合适的时机或者标记来控制删除，否则Unmount的执行比Memo的执行晚， 会在后面把任务移除掉的
        // context.removeTask(getId(props));
      };
    }, [props.getId()]);
  }

  /**
   * 绑定rdxNode到context中, 并绑定状态了状态
   * @param node
   */
  function useRdxNodeBinding<GModel>(
    node: RdxNode<GModel> | string
  ): DataContext<GModel> {
    const context = useRdxStateContext(provider);
    if (node instanceof RdxNode) {
      useTaskBinding(node);
    } else if (!context.hasTask(node)) {
      throw new Error(`未找到id为${node}的节点`);
    }
    useStateUpdate(getDepId(node), StateUpdateType.State);

    const data: DataContext<GModel> = {
      ...createBaseContext(getDepId(node), context),
      ...createMutators(getDepId(node), context),
    };
    return data;
  }

  // ?当依赖的组件移除了怎么办，为啥共享组件数据的时候，数据需要提到父级， 因为当组件移除的情况下，父级组件用户可以主动的控制
  function useRdxAtom<GModel>(props: IRdxAtomNode<GModel>) {
    const rdxAtom = new RdxAtomNode(props);
    const dataContext = useRdxNodeBinding(rdxAtom);
    return [dataContext.value, dataContext.next, rdxAtom] as [
      GModel,
      (v: GModel) => void,
      RdxAtomNode<GModel>
    ];
  }
  function useRdxAtomLoader<GModel>(
    props: IRdxAtomNode<GModel>
  ): [GModel, TNext<GModel>, DataContext<GModel>, RdxAtomNode<GModel>] {
    const rdxAtom = new RdxAtomNode(props);
    const dataContext = useRdxNodeBinding<GModel>(rdxAtom);
    useStateUpdate(rdxAtom.getId(), StateUpdateType.ReactionStatus);
    return [dataContext.value, dataContext.next, dataContext, rdxAtom];
  }

  function useRdxRefrence<GModel>(id: string) {
    return useRdxNodeBinding<GModel>(id);
  }
  function useRdxWatcher<GModel>(props: IRdxWatcherNode<GModel>, deps?: any[]) {
    const newProps = { ...props };
    // 这里通过rdxWatcher实现类比WatcherFamily的功能
    if (deps && Array.isArray(deps)) {
      newProps.id = `${newProps.id}__watcher/${JSON.stringify(deps)}`;
    }
    const dataContext = useRdxNodeBinding(new RdxWatcherNode(newProps));
    return [dataContext.value, dataContext.next] as [GModel, TNext<GModel>];
  }

  function useRdxWatcherLoader<GModel>(
    props: IRdxWatcherNode<GModel>,
    deps?: string[]
  ): [GModel, TNext<GModel>, DataContext<GModel>] {
    const newProps = { ...props };
    // 这里通过rdxWatcher实现类比WatcherFamily的功能
    if (deps && Array.isArray(deps)) {
      newProps.id = `${newProps.id}__watcher/${JSON.stringify(deps)}`;
    }
    const watcher = new RdxWatcherNode(newProps);
    useStateUpdate(watcher.getId(), StateUpdateType.ReactionStatus);
    const dataContext = useRdxNodeBinding<GModel>(watcher);
    return [dataContext.value, dataContext.next, dataContext];
  }
  function useRdxGlobalContext() {
    const context = useRdxStateContext(provider);
    useGlobalStateUpdate(context);
    return {
      taskState: context.getAllTaskState(),
      virtualTaskState: context.getAllVirtualTaskState(),
    };
  }
  function useRdxGlboalState() {
    const context = useRdxStateContext(provider);
    useGlobalStateUpdate(context);
    return context.getAllTaskState();
  }
  // =======================================基础hooks=========================================================
  function useRdxState<GModel>(
    props: RdxNode<GModel> | string
  ): [GModel, (v: GModel) => void] {
    const context = useRdxNodeBinding<GModel>(props);
    return [context.value, context.next];
  }

  function useRdxSetter<GModel>(props: RdxNode<GModel> | string) {
    const context = useRdxNodeBinding<GModel>(props);
    return context.next;
  }
  function useRdxStateLoader<GModel>(
    node: RdxNode<GModel> | string
  ): [Status, GModel, (v: GModel) => void] {
    const context = useRdxNodeBinding<GModel>(node);
    useStateUpdate(getDepId(node), StateUpdateType.ReactionStatus);
    return [context.status, context.value, context.next];
  }

  function useRdxValue<GModel>(node: RdxNode<GModel> | string): GModel {
    let context = useRdxNodeBinding<GModel>(node);
    return context.value;
  }

  function useRdxValueLoader<GModel>(node: RdxNode<GModel>): [Status, any] {
    const context = useRdxNodeBinding(node);
    useStateUpdate(getDepId(node), StateUpdateType.ReactionStatus);
    return [context.status, context.value];
  }

  // =======================================基础hooks=========================================================

  let reactionPreviewId = 0;
  function useRdxValueByDependencies<IRelyModel>(props: {
    displayName?: string;
    deps: IRdxAnyDeps[];
  }): IRelyModel {
    const { deps = [], displayName } = props;
    const uniqueId = useRef(`${displayName}-preview` + reactionPreviewId++);
    const context = useRdxWatcher({
      id: uniqueId.current,
      get: ({ get }) => {
        return deps.map((dep) => get(dep));
      },
    });
    return (context[0] as unknown) as any;
  }
  return {
    useRdxAtom,
    useRdxAtomLoader,
    useRdxWatcher,
    useRdxWatcherLoader,
    useRdxGlboalState,
    useRdxGlobalContext,
    useRdxRefrence,
    useRdxState,
    useRdxValue,
    useRdxStateLoader,
    useRdxValueLoader,
    useRdxValueByDependencies,
    useRdxSetter,
  };
}

const defaultHooks = createRdxHooks();
export const useRdxAtom = defaultHooks.useRdxAtom;
export const useRdxAtomLoader = defaultHooks.useRdxAtomLoader;
export const useRdxWatcher = defaultHooks.useRdxWatcher;
export const useRdxWatcherLoader = defaultHooks.useRdxWatcherLoader;
export const useRdxGlboalState = defaultHooks.useRdxGlboalState;
export const useRdxRefrence = defaultHooks.useRdxRefrence;
export const useRdxState = defaultHooks.useRdxState;
export const useRdxValue = defaultHooks.useRdxValue;
export const useRdxStateLoader = defaultHooks.useRdxStateLoader;
export const useRdxValueLoader = defaultHooks.useRdxValueLoader;
export const useRdxValueByDependencies = defaultHooks.useRdxValueByDependencies;
export const useRdxSetter = defaultHooks.useRdxSetter;
