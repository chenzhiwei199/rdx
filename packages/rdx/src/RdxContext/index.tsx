import * as React from 'react';
import { initValue, ShareContextClass } from './shareContext';
import { RdxContextProps } from './interface';
import { ScopeObject } from './core';
import { StateUpdateType } from '../global';
import { DefaultContext } from '../hooks/stateHooks';
import { DataPersistenceHook, TaskEventTriggerType } from '../DataPersist';
export * from './core';
export * from './interface';
const Rdx = (props: RdxContextProps) => {
  const {
    initializeState = {},
    onChange = () => {},
    onLoading = () => {},
    name,
    // withRef,
    context = DefaultContext ,
    createStore,
    visualStatePlugins,
  } = props;
  // 如果initializeState为undefined, 则需要替换状态， 否则不需要
  // 创建store
  function createTaskState(value: any) {
    return createStore
      ? createStore(initializeState)
      : new ScopeObject(initializeState);
  }

  // context下保留不变的store
  const store = React.useRef(
    new ShareContextClass({
      ...initValue(),
      name,
      virtualTaskState: createTaskState(initializeState),
      taskState: createTaskState(initializeState),
    })
  );

  // 绑定批量更新方法
  store.current.setChangeCallback((v) => {
    store.current.getEventEmitter().emit(StateUpdateType.GlobalState);
    onChange(v);
  });
  store.current.setLoadingCallback(() => {
    onLoading()
  })

  // withRef && (withRef.current = store.current);
  let copyState = store.current.getAllTaskState();
  React.useEffect(() => {
    store.current.parentMounted = true;
    // 执行初始化任务
    store.current.executeTask(
      Array.from(store.current.getNotifyQueue()),
      TaskEventTriggerType.BatchReactionOnMount
    );
    store.current.getNotifyQueue().clear();

    // 初始化状态和后续状态不一样，则触发onChange
    if (copyState !== store.current.getAllTaskState() && !store.current.taskScheduler.isRunning()) {
      onChange(store.current.getAllTaskState());
    }
  }, []);
  return (
    <context.Provider value={store.current}>
      <DevTool shareContext={context}/>
      {visualStatePlugins}
      {props.children}
    </context.Provider>
  );
};

const DevTool = (props: { shareContext?: React.Context<ShareContextClass>}) => {

  DataPersistenceHook(props.shareContext)
  return <></>
}
export const RdxContext = Rdx;
