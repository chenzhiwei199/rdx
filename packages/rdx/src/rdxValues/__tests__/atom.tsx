
import renderer from 'react-test-renderer';
import {render} from '@testing-library/react';
import { RdxContext } from '../../RdxContext';
import React from 'react';
import { useRdxAtom } from '../../hooks/stateHooks';

test('当Rdx外部初始化了数据，不应该被默认值设置覆盖', () => {
  const Test = () => {
    const [atom] = useRdxAtom({
      id: 'test',
      defaultValue: 'default'
    })
    return <div>{atom}</div>
  }
  const { getByText } = render(
    <RdxContext initializeState={{ 'test': 'init'}}>
      <Test></Test>
    </RdxContext>
  );
  expect(getByText('init')).toBeTruthy();
});