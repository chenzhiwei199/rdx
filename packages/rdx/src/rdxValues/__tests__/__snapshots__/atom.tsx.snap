// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`当Rdx外部初始化了数据，不应该被默认值设置覆盖 1`] = `
<div>
  init
</div>
`;
